src/dataflow/src/source/kafka.rs

```diff
@@ -624,29 +665,41 @@ struct PartitionConsumer {
     pid: i32,
     /// The underlying Kafka partition queue
     partition_queue: PartitionQueue<GlueConsumerContext>,
+    /// Whether or not to unpack and allocate headers and pass them through in the `SourceMessage`
+    include_headers: bool,
 }
 
 impl PartitionConsumer {
     /// Creates a new partition consumer from underlying Kafka consumer
-    fn new(pid: i32, partition_queue: PartitionQueue<GlueConsumerContext>) -> Self {
+    fn new(
+        pid: i32,
+        partition_queue: PartitionQueue<GlueConsumerContext>,
+        include_headers: bool,
+    ) -> Self {
         PartitionConsumer {
             pid,
             partition_queue,
+            include_headers,
         }
     }
 
     /// Returns the next message to process for this partition (if any).
+    /// The outer `Result` represents irrecoverable failures, the inner one can and will
+    /// be transformed into empty values
     fn get_next_message(
         &mut self,
-    ) -> Result<Option<SourceMessage<Option<Vec<u8>>, Option<Vec<u8>>>>, KafkaError> {
+    ) -> Result<
+        Result<Option<SourceMessage<Option<Vec<u8>>, Option<Vec<u8>>>>, KafkaError>,
+        anyhow::Error,
+    > {
```

Comment:
This return type is grim. I think we need to find a way to make this less painful. Not sure what, but triply nested `Result`s/`Option`s are beyond my threshold for code clarity!

Reply:
I added a `Headers::iter` method that is nonfallible that you can use in `construct_headers`, which should allow you to elide this outer `Result`.

Reply:
I can definitely make this into some type-aliases, and use the new methods!

Reply:
I think

```
Result<Option<SourceMessage<Option<Vec<u8>>, Option<Vec<u8>>>>, KafkaError>
```

is workable enough if you want to leave as is! Would be good to get a typedef like `type SourceMessageBytes = SourceMessage<Option<Vec<u8>>, Option<Vec<u8>>>`, but no need to expand the scope of this PR if it does turn out you can leave the signature alone now!